<!DOCTYPE html>
<!-- saved from url=(0014)about:internet -->
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Metagenome R workflow version 1.1</title>

<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 12px;
   margin: 8px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 { 
   font-size:2.2em; 
}

h2 { 
   font-size:1.8em; 
}

h3 { 
   font-size:1.4em; 
}

h4 { 
   font-size:1.0em; 
}

h5 { 
   font-size:0.9em; 
}

h6 { 
   font-size:0.8em; 
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre {	
   margin-top: 0;
   max-width: 95%;
   border: 1px solid #ccc;
   white-space: pre-wrap;
}

pre code {
   display: block; padding: 0.5em;
}

code.r, code.cpp {
   background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * { 
      background: transparent !important; 
      color: black !important; 
      filter:none !important; 
      -ms-filter: none !important; 
   }

   body { 
      font-size:12pt; 
      max-width:100%; 
   }
       
   a, a:visited { 
      text-decoration: underline; 
   }

   hr { 
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote { 
      padding-right: 1em; 
      page-break-inside: avoid; 
   }

   tr, img { 
      page-break-inside: avoid; 
   }

   img { 
      max-width: 100% !important; 
   }

   @page :left { 
      margin: 15mm 20mm 15mm 10mm; 
   }
     
   @page :right { 
      margin: 15mm 10mm 15mm 20mm; 
   }

   p, h2, h3 { 
      orphans: 3; widows: 3; 
   }

   h2, h3 { 
      page-break-after: avoid; 
   }
}

</style>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: rgb(88, 72, 246)
   }

   pre .number {
     color: rgb(0, 0, 205);
   }

   pre .comment {
     color: rgb(76, 136, 107);
   }

   pre .keyword {
     color: rgb(0, 0, 255);
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: rgb(3, 106, 7);
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>




</head>

<body>
<h1>Metagenome R workflow version 1.1</h1>

<h2>Introduction</h2>

<p>This R tutorial show how to extract individual genome bins from metagenomes, as descibed in Albertsen et al., 2013 (Add link). The guide is written in R markdown and can be found <a href="https://github.com/MadsAlbertsen/multi-metagenome/tree/master/R.data.generation">HERE</a>.</p>

<p>The basic requirement is two metagenomes where the target species are in differential abundance. The data is assembled into 1 assembly (i.e a collection of scaffolds). The raw reads are then mapped independently to the assembly, which generates two coverage estimates for each scaffold.</p>

<p>The guide uses the original data from the publication.</p>

<hr/>

<h3>Download the R formatted data.</h3>

<p>Download and upack the data used in Albertsen et al., 2013. The file is approximately 70 Mb.</p>

<pre><code class="r">#
# download.file(&#39;https://dl.dropbox.com/s/989dix16ugyuvrq/Albertsen2013.data.tar.gz&#39;,
# &#39;Albertsen2013.data.tar.gz&#39;) untar(&#39;Albertsen2013.data.tar.gz&#39;)
</code></pre>

<hr/>

<h3>Load needed packages</h3>

<p>In case you havn&#39;t install all the needed packages, they can be installed via e.g. <strong>install.packages(&#39;vegan&#39;)</strong>.</p>

<pre><code class="r">R.version$version.string
</code></pre>

<pre><code>## [1] &quot;R version 3.0.0 (2013-04-03)&quot;
</code></pre>

<pre><code class="r">library(&quot;vegan&quot;)
library(&quot;plyr&quot;)
library(RColorBrewer)
library(alphahull)
library(ggplot2)
</code></pre>

<hr/>

<h3>Read in all the data</h3>

<p>All data except the two coverage estimates (<strong>HPminus</strong>, <strong>HPplus</strong>) can be automatically generated from a fasta file of the assembled scaffolds using the script: <strong>workflow.R.data.generation.sh</strong>. Coverage estimates for the scaffolds can be obtained through most short read mappers.</p>

<pre><code class="r">HPminus &lt;- read.csv(&quot;HPminus.scaffold.coverage.csv&quot;, header = T)
HPplus &lt;- read.csv(&quot;HPplus.scaffold.coverage.csv&quot;, header = T)
gc &lt;- read.delim(&quot;assembly.gc.tab&quot;, header = T)
kmer &lt;- read.delim(&quot;assembly.kmer.tab&quot;, header = T)
ess &lt;- read.table(&quot;assembly.orfs.hmm.id.txt&quot;, header = F)
ess.tax &lt;- read.delim(&quot;assembly.orfs.hmm.blast.tax.tab&quot;, header = F)
cons.tax &lt;- read.delim(&quot;assembly.tax.consensus.txt&quot;, header = T)

colnames(kmer)[1] = &quot;name&quot;
colnames(ess) = c(&quot;name&quot;, &quot;orf&quot;, &quot;hmm.id&quot;)
colnames(ess.tax) = c(&quot;name&quot;, &quot;orf&quot;, &quot;phylum&quot;)
colnames(cons.tax) = c(&quot;name&quot;, &quot;phylum&quot;, &quot;tax.color&quot;, &quot;all.assignments&quot;)
</code></pre>

<p>Merge all data on scaffolds into a single data frame <strong>d</strong>.</p>

<pre><code class="r">d &lt;- as.data.frame(cbind(HPminus$Name, HPplus$Reference.length, gc$gc, HPminus$Average.coverage, 
    HPplus$Average.coverage), row.names = F)
colnames(d) = c(&quot;name&quot;, &quot;length&quot;, &quot;gc&quot;, &quot;HPminus&quot;, &quot;HPplus&quot;)
d &lt;- merge(d, cons.tax, by = &quot;name&quot;, all = T)
</code></pre>

<p>Clean up phylum names.</p>

<pre><code class="r">d$phylum &lt;- sub(&quot;&lt;phylum&gt;&quot;, &quot;&quot;, d$phylum)
d$phylum &lt;- sub(&quot;unclassified Bacteria&quot;, &quot;TM7&quot;, d$phylum)
d$phylum &lt;- sub(&quot;/Chlorobi group&quot;, &quot;&quot;, d$phylum)
d$phylum &lt;- sub(&quot;Chlamydiae/&quot;, &quot;&quot;, d$phylum)
d$phylum &lt;- sub(&quot; group&quot;, &quot;&quot;, d$phylum)
d$phylum &lt;- sub(&quot;Amoebozoa&quot;, NA, d$phylum)
d$phylum &lt;- sub(&quot;Opisthokonta&quot;, NA, d$phylum)
</code></pre>

<p>Merge all data on essential genes into a single data frame <strong>e</strong>.</p>

<pre><code class="r">e &lt;- merge(ess, d, by = &quot;name&quot;, all.x = T)
e &lt;- merge(e, ess.tax, by = c(&quot;name&quot;, &quot;orf&quot;), all.x = T)
</code></pre>

<hr/>

<h3>Define a few functions for later use</h3>

<p>Calculate basic statistics on a set of scaffolds.</p>

<pre><code class="r">genome.stats &lt;- matrix(NA, nrow = 0, ncol = 9)
colnames(genome.stats) &lt;- c(&quot;total.length&quot;, &quot;# scaffolds&quot;, &quot;mean.length&quot;, &quot;max.length&quot;, 
    &quot;gc&quot;, &quot;HPminus&quot;, &quot;HPplus&quot;, &quot;tot.ess&quot;, &quot;uni.ess&quot;)

calc.genome.stats &lt;- function(x, y) matrix(c(sum(x$length), nrow(x), round(mean(x$length), 
    1), max(x$length), round(sum((x$gc * x$length))/sum(x$length), 1), round(sum((x$HPminus * 
    x$length))/sum(x$length), 1), round(sum((x$HPplus * x$length))/sum(x$length), 
    1), nrow(y), length(unique(y$hmm.id))), dimnames = list(colnames(genome.stats), 
    &quot;&quot;))
</code></pre>

<p>Extract a subset of scaffolds.</p>

<pre><code class="r">extract &lt;- function(x, a.def, v1, v2) {
    out &lt;- {
    }
    for (i in 1:nrow(x)) {
        if (inahull(a.def, c(v1[i], v2[i]))) 
            out &lt;- rbind(out, x[i, ])
    }
    return(out)
}
</code></pre>

<hr/>

<h2>Initial overview of the data</h2>

<pre><code class="r">calc.genome.stats(d, e)
</code></pre>

<pre><code>##                       
## total.length 4.231e+08
## # scaffolds  1.339e+05
## mean.length  3.159e+03
## max.length   3.049e+06
## gc           5.460e+01
## HPminus      8.270e+01
## HPplus       2.460e+01
## tot.ess      8.311e+03
## uni.ess      1.080e+02
</code></pre>

<p>To get an initial overview of the data we only use scaffolds &gt; 5000 bp.</p>

<pre><code class="r">ds &lt;- subset(d, length &gt; 5000)
es &lt;- subset(e, length &gt; 5000)
</code></pre>

<h3>Coverage plots - Colored by GC</h3>

<pre><code class="r">ggplot(ds, aes(x = HPminus, y = HPplus, color = gc, size = length)) + scale_x_log10(limits = c(5, 
    5000)) + scale_y_log10(limits = c(0.01, 2000)) + xlab(&quot;Coverage (HP-)&quot;) + 
    ylab(&quot;Coverage (HP+)&quot;) + geom_point(alpha = 0.5) + scale_size_area(name = &quot;Scaffold length&quot;, 
    max_size = 20) + scale_colour_gradientn(colours = c(&quot;red&quot;, &quot;green&quot;, &quot;blue&quot;))
</code></pre>

<p><img src="figure/Overview%20-%20Coverage%20GC%20plot.png" alt="plot of chunk Overview - Coverage GC plot"/> </p>

<h3>Coverage plots - Colored by phylum level assignment of essential genes</h3>

<p>Use the tax.color variable to only color scaffolds from the 7 most abundant phyla. The tax.color variable is sorted by abundance. E.g. tax.color = 1 is assigned to the phyla with most scaffolds assigned. Change the <strong>t</strong> parameter to include more or less phyla.</p>

<pre><code class="r">t &lt;- 8
ds$tax.color[is.na(ds$tax.color)] &lt;- 0
for (i in 1:nrow(ds)) {
    if (as.integer(ds$tax.color[i]) &lt; t &amp; as.integer(ds$tax.color[i]) &gt; 0) {
        ds$tax.color[i] &lt;- brewer.pal(8, &quot;Paired&quot;)[as.integer(ds$tax.color[i])]
    } else {
        ds$tax.color[i] &lt;- NA
        ds$phylum[i] &lt;- NA
    }
}

pcol &lt;- cbind(unique(ds$tax.color)[-1], unique(ds$phylum)[-1])
pcol &lt;- pcol[order(pcol[, 2]), 1]
</code></pre>

<pre><code class="r">ggplot(ds, aes(x = HPminus, y = HPplus, size = length, colour = phylum)) + scale_x_log10(limits = c(5, 
    5000)) + scale_y_log10(limits = c(0.01, 2000)) + xlab(&quot;Coverage (HP-)&quot;) + 
    ylab(&quot;Coverage (HP+)&quot;) + geom_point(alpha = 0.1, colour = &quot;black&quot;) + geom_point(shape = 1) + 
    scale_colour_manual(name = &quot;Phyla&quot;, values = pcol) + scale_size_area(name = &quot;Scaffold length&quot;, 
    max_size = 20) + guides(colour = guide_legend(override.aes = list(alpha = 1, 
    size = 5, shape = 19)))
</code></pre>

<p><img src="figure/Overview%20-%20Coverage%20phylum%20plot.png" alt="plot of chunk Overview - Coverage phylum plot"/> </p>

<hr/>

<h2>Genome extraction</h2>

<p>First zoom on a target genome - here the Verrumicrobia. Use the scaffolds with essential genes as a rough guide for selection of a subset of scaffolds that include the target genome. The non-target scaffolds will be removed in the next step. The locater function is used to interactively define a subspace on the plot. As locater is interactive - I&#39;ve added the points maunally to allow recration of the full guide. The area defined by the selected points is extracted using the ahull function. </p>

<pre><code class="r">x &lt;- &quot;HPminus&quot;
y &lt;- &quot;HPplus&quot;

plot(ds[, x], ds[, y], log = &quot;xy&quot;, cex = sqrt(ds$length)/100, pch = 20, col = rgb(0, 
    0, 0, 0.1), xlim = c(55, 110), ylim = c(0.5, 10), xlab = &quot;Coverage HP-&quot;, 
    ylab = &quot;Coverage HP+&quot;)

points(ds[, x], ds[, y], cex = sqrt(ds$length)/100 * 0.7, col = ds$tax.color, 
    lwd = 2)

# def&lt;-locator(100, type=&#39;p&#39;, pch=20)

def &lt;- {
}
def$x &lt;- c(64, 66, 81, 92, 94, 81, 68, 65)
def$y &lt;- c(2, 6.6, 7.7, 3.9, 1.4, 1, 1, 1.4)

g1.selection.A &lt;- ahull(def, alpha = 1e+05)

plot(g1.selection.A, col = &quot;black&quot;, add = T)
</code></pre>

<p><img src="figure/GE%20-%20Zoom.png" alt="plot of chunk GE - Zoom"/> </p>

<p>Extract all scaffolds and information on essential genes within the defined subspace using the <strong>extract</strong> function.</p>

<pre><code class="r">g1.s.A &lt;- extract(ds, g1.selection.A, ds[, x], ds[, y])
g1.e.A &lt;- extract(es, g1.selection.A, es[, x], es[, y])
</code></pre>

<p>See the basic statistics of the selected scaffolds.</p>

<pre><code class="r">calc.genome.stats(g1.s.A, g1.e.A)
</code></pre>

<pre><code>##                       
## total.length 9801161.0
## # scaffolds      228.0
## mean.length    42987.5
## max.length    701203.0
## gc                54.9
## HPminus           79.1
## HPplus             3.2
## tot.ess          161.0
## uni.ess          104.0
</code></pre>

<p>Which of the single copy genes are duplicated? Note that some genomes might have duplicates of some &ldquo;single copy genes&rdquo;.</p>

<pre><code class="r">g1.d.A &lt;- g1.e.A[which(duplicated(g1.e.A$hmm.id) | duplicated(g1.e.A$hmm.id, 
    fromLast = TRUE)), ]
head(g1.d.A[order(g1.d.A$hmm.id), c(1, 3, 8)])
</code></pre>

<pre><code>##       name     hmm.id        phylum.x
## 2123  1820 PF00162.14 Verrucomicrobia
## 4599  3996 PF00162.14             TM7
## 1322  1328 PF00380.14 Verrucomicrobia
## 5090  4623 PF00380.14             TM7
## 1611 14624 PF00411.14 Verrucomicrobia
## 5095  4623 PF00411.14             TM7
</code></pre>

<p>As there is multiple genomes in the subset we make a PCA on the scaffolds in the subset.</p>

<pre><code class="r">rda &lt;- rda(kmer[g1.s.A$name, 2:ncol(kmer)], scale = T)
scores &lt;- scores(rda, choices = 1:5)$sites

g1.s.B &lt;- cbind(g1.s.A, scores)
g1.e.B &lt;- merge(g1.e.A, g1.s.B[, c(1, 9:13)], all.x = T, by = &quot;name&quot;)
g1.d.B &lt;- merge(g1.d.A, g1.s.B[, c(1, 9:13)], all.x = T, by = &quot;name&quot;)
</code></pre>

<p>We use the pairs function to plot the first 5 components.</p>

<pre><code class="r">rgb.c &lt;- colorRampPalette(c(&quot;red&quot;, &quot;green&quot;, &quot;blue&quot;))
rgb.a &lt;- adjustcolor(rgb.c(max(d$gc) - min(d$gc)), alpha.f = 0.2)
palette(rgb.a)

pairs(g1.s.B[, 9:13], upper.panel = NULL, col = g1.s.B$gc - min(d$gc), cex = sqrt(g1.s.B$length)/100, 
    pch = 20)
</code></pre>

<p><img src="figure/GE%20-%20plot%20PCA.png" alt="plot of chunk GE - plot PCA"/> </p>

<p>PC1 and PC2 seem to seperate our target genome from the other scaffolds and is therefore used for another extraction using the locator function.</p>

<pre><code class="r">x &lt;- &quot;PC1&quot;
y &lt;- &quot;PC2&quot;

plot(g1.s.B[, x], g1.s.B[, y], cex = sqrt(g1.s.B$length)/100, pch = 20, col = rgb(0, 
    0, 0, 0.1), xlab = x, ylab = y)

points(g1.s.B[, x], g1.s.B[, y], cex = sqrt(g1.s.B$length)/100 * 0.7, col = g1.s.B$tax.color, 
    lwd = 1)

# def&lt;-locator(100, type=&#39;p&#39;, pch=20)

def &lt;- {
}
def$x &lt;- c(0.3740306, 0.4839196, 0.9084907, 1.2431527, 1.2781173, 1.0733242, 
    0.653748, 0.4689347, 0.3690356)
def$y &lt;- c(0.2810738, 1.31294166, 1.94015545, 1.99073721, 1.33317436, 0.39235367, 
    0.04839772, 0.02816501, 0.22037569)

g1.selection.B &lt;- ahull(def, alpha = 1e+05)

plot(g1.selection.B, col = &quot;black&quot;, add = T)
</code></pre>

<p><img src="figure/GE%20-%20Extract%20PCA.png" alt="plot of chunk GE - Extract PCA"/> </p>

<p>Again the <strong>extract</strong> function is used to retrive the scaffolds in the selected subset.</p>

<pre><code class="r">g1.s.C &lt;- extract(g1.s.B, g1.selection.B, g1.s.B[, x], g1.s.B[, y])
g1.e.C &lt;- extract(g1.e.B, g1.selection.B, g1.e.B[, x], g1.e.B[, y])
</code></pre>

<p>See the basic statistics of the selected scaffolds.</p>

<pre><code class="r">calc.genome.stats(g1.s.C, g1.e.C)
</code></pre>

<pre><code>##                       
## total.length 6823444.0
## # scaffolds       59.0
## mean.length   115651.6
## max.length    701203.0
## gc                60.9
## HPminus           77.4
## HPplus             3.0
## tot.ess          106.0
## uni.ess          103.0
</code></pre>

<p>There are a few duplicated &ldquo;single copy genes&rdquo; however in this case it is not due to mulitple species in the bin, but real duplicates in the genome.</p>

<pre><code class="r">g1.d.C &lt;- g1.e.C[which(duplicated(g1.e.C$hmm.id) | duplicated(g1.e.C$hmm.id, 
    fromLast = TRUE)), ]
head(g1.d.C[order(g1.d.C$hmm.id), c(1, 3, 8)])
</code></pre>

<pre><code>##     name     hmm.id        phylum.x
## 42   735 PF01795.14 Verrucomicrobia
## 79  2546 PF01795.14 Verrucomicrobia
## 80  2546 PF01795.14 Verrucomicrobia
## 139 5388 PF01795.14 Verrucomicrobia
</code></pre>

<p>Add the genome statistics to a list and print the name of the scaffolds to a file for further refinement.</p>

<pre><code class="r">genome.stats &lt;- rbind(genome.stats, t(calc.genome.stats(g1.s.C, g1.e.C)))
rownames(genome.stats)[nrow(genome.stats)] &lt;- &quot;genome 1&quot;
show(genome.stats)
</code></pre>

<pre><code>##          total.length # scaffolds mean.length max.length   gc HPminus
## genome 1      6823444          59      115652     701203 60.9    77.4
##          HPplus tot.ess uni.ess
## genome 1      3     106     103
</code></pre>

<pre><code class="r">write.table(g1.s.C$name, file = &quot;genome1.txt&quot;, quote = F, row.names = F, col.names = F)
</code></pre>

</body>

</html>

